src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,169 @@
+            tail = head;
+            size++;
+        } else if (index == 0) {
+            Node<T> nodeToAdd = new Node<T>(null, value, tail);
+            tail.prev = nodeToAdd;
+            tail = nodeToAdd;
+            size++;
+        } else if (index == size) {
+            add(value);
+        } else {
+            Node<T> currentNode = getNode(index);
+            Node<T> nodeToAdd = new Node<T>(currentNode.prev, value, currentNode);
+            currentNode.prev.next = nodeToAdd;
+            currentNode.prev = nodeToAdd;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T el : list) {
+            add(el);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        return getNode(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        Node<T> currentNode = getNode(index);
+        Node<T> nodeToSet = new Node<T>(currentNode.prev, value, currentNode.next);
+        if (currentNode.next != null) {
+            currentNode.next.prev = nodeToSet;
+        } else {
+            head = nodeToSet;
+        }
+        if (currentNode.prev != null) {
+            currentNode.prev.next = nodeToSet;
+        } else {
+            tail = nodeToSet;
+        }
+        return currentNode.item;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        Node<T> currentNode = getNode(index);
+
+        remove(currentNode.item);
+        return currentNode.item;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> currentNode = getNode(object);
+        if (currentNode == null) {
+            return false;
+        }
+        unlink(currentNode);
+        size--;
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return tail == null && head == null;
+    }
+
+    private Node<T> getNode(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+
+        Node<T> foundedNode = tail;
+        for (int i = 0; i < index; i++) {
+            if (foundedNode == null) {
+                throw new IndexOutOfBoundsException();
+            }
+            foundedNode = foundedNode.next;
+        }
+        return foundedNode;
+    }
+
+    private Node<T> getNode(T value) {
+        Node<T> currentNode = tail;
+        while (!Objects.equals(value, currentNode.item)) {
```

Comment:
we still can't use Objects.equals (we will be able to in HashMap HW), so write your own code here